### PR TITLE
docs: add `CONTRIBUTING.md`

### PR DESCRIPTION
**Pull request by bhattabhi013**
## Description

Added `CONTRIBUTING.md` in the project.

Fixes issue #394 

## Type of Change

- [ ] ✨ `feat` -- New feature (non-breaking change which adds functionality)
- [ ] 🛠️ `fix` -- Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ `!` -- Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 `refactor` -- Code refactor
- [ ] ✅ `ci` -- Build configuration change
- [x] 📝 `docs` -- Documentation
- [ ] 🗑️ `chore` -- Chore

### CODE REVIEW

**Comment by bhattabhi013** [comment]
@blaugold, please review the PR.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+_See also: [Flutter's code of conduct](https://flutter.dev/design-principles/#code-of-conduct)_
+
+## About
+
+Melos is a tool that optimizes the workflow around managing multi-package repositories with git and Pub.
+
+
+## Contribution types
+
```
Comment (nilsreichardt) [inline CONTRIBUTING.md:14]:
```suggestion
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+### Bug Report
+
+- If you find a bug, please first report it using [Github issues].
+  - First check if there is not already an issue for it; duplicated issues will be closed.
+
```
Comment (nilsreichardt) [inline CONTRIBUTING.md:20]:
```suggestion
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+### Bug Fix
+
+- If you'd like to submit a fix for a bug, please read the [How To](#how-to-contribute) for how to send a Pull Request.
+- Indicate on the open issue that you are working on fixing the bug and the issue will be assigned to you.
+- Write `Fix #xxxx` in your PR text, where xxxx is the issue number (if there is one).
+- Include a test that isolates the bug and verifies that it was fixed.
+
```
Comment (nilsreichardt) [inline CONTRIBUTING.md:28]:
```suggestion
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+### Documentation & Miscellaneous
+
+- If you have suggestions for improvements to the documentation or examples (or something else), we would love to hear about it.
+- As always first file a [Github issue].
+- Implement the changes to the documentation, please read the [How To](#how-to-contribute).
+
+## How To Contribute
+
```
Comment (nilsreichardt) [inline CONTRIBUTING.md:45]:
```suggestion
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+- Include a test that isolates the bug and verifies that it was fixed.
+
+
+### New Features
+
+- If you'd like to add a feature to the library that doesn't already exist, feel free to describe the feature in a new [GitHub issue].
+- You can also join us on [Discord]() to discuss some initials thoughts.
+- If you'd like to implement the new feature, please wait for feedback from the project maintainers before spending too much time writing the code. In some cases, enhancements may not align well with the project future development direction.
+- If applicable, implement the code for the new feature and please read the [How To](#how-to-contribute).
+
```
Comment (nilsreichardt) [inline CONTRIBUTING.md:36]:
```suggestion
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+- Ensure all the dependencies described in the previous section are installed.
+- Fork `https://github.com/invertase/melos` into your own GitHub account. If
+  you already have a fork, and are now installing a development environment on
+  a new machine, make sure you've updated your fork so that you don't use stale
+  configuration options from long ago.
+- `git clone git@github.com:<your_name_here>/melos.git`
+- `git remote add upstream git@github.com:invertase/melos.git` (So that you
+  fetch from the main repository, not your clone, when running `git fetch` or `git pull`
+
```
Comment (nilsreichardt) [inline CONTRIBUTING.md:62]:
```suggestion
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+- If you have suggestions for improvements to the documentation or examples (or something else), we would love to hear about it.
+- As always first file a [Github issue].
+- Implement the changes to the documentation, please read the [How To](#how-to-contribute).
+
+## How To Contribute
+
+
+### Requirements
+- Linux, Mac OS X, or Windows.
```
Comment (nilsreichardt) [inline CONTRIBUTING.md:48]:
```suggestion

- Linux, Mac OS X, or Windows.
```

**Review by nilsreichardt — COMMENTED** [review]
Just to be consistent about headline style 👍 _Note:_ I have not reviewed the content.

File: CONTRIBUTING.md (hunk 1 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:13]: Done

File: CONTRIBUTING.md (hunk 2 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:18]: Done

File: CONTRIBUTING.md (hunk 3 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:25]: Done

File: CONTRIBUTING.md (hunk 4 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:40]: Done

File: CONTRIBUTING.md (hunk 5 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:32]: Done

File: CONTRIBUTING.md (hunk 6 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:57]: Done

File: CONTRIBUTING.md (hunk 7 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:43]: Done

**Comment by bhattabhi013** [comment]
I've tried incorporating the suggestions and fixing the issues.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
+## Contribution Guidelines
```
Comment (blaugold) [inline CONTRIBUTING.md:1]: This should be a level 1 heading.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+**Note:** If these contribution guidelines are not followed your issue or PR might be closed, so
+please read these instructions carefully.
+
+_See also: [Flutter's code of conduct](https://flutter.dev/design-principles/#code-of-conduct)_
```
Comment (blaugold) [inline CONTRIBUTING.md:6]: This is Flutter's code of conduct. Not sure if invertase has their own or if we should include Flutter's code of conduct.

cc @Salakar

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+## About
+
+Melos is a tool that optimizes the workflow around managing multi-package repositories with git and Pub.
+
+## Contribution types
+
+### Bug Report
+
+- If you find a bug, please first report it using [Github issues].
+  - First check if there is not already an issue for it; duplicated issues will be closed.
```
Comment (blaugold) [inline CONTRIBUTING.md:17]: Is this item meant to be indented?

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+### Bug Fix
+
+- If you'd like to submit a fix for a bug, please read the [How To](#how-to-contribute) for how to send a Pull Request.
+- Indicate on the open issue that you are working on fixing the bug and the issue will be assigned to you.
+- Write `Fix #xxxx` in your PR text, where xxxx is the issue number (if there is one).
+- Include a test that isolates the bug and verifies that it was fixed.
+
+### New Features
+
+- If you'd like to add a feature to the library that doesn't already exist, feel free to describe the feature in a new [GitHub issue].
```
Comment (blaugold) [inline CONTRIBUTING.md:28]: There is no URL defined to link `[GitHub issue]`?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+- You can also join us on [Discord]() to discuss some initials thoughts.
```
Comment (blaugold) [inline CONTRIBUTING.md:29]: We don't have a discord :)

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+# ---------------------------------------------------------
+# | You are running a local development version of melos. |
+# ---------------------------------------------------------
+```
+
+### Install
+
+```bash
+dart pub global activate melos
+```
```
Comment (blaugold) [inline CONTRIBUTING.md:83]: This section is not necessary.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+- Ensure all the dependencies described in the previous section are installed.
+- Fork `https://github.com/invertase/melos` into your own GitHub account. If
+  you already have a fork, and are now installing a development environment on
+  a new machine, make sure you've updated your fork so that you don't use stale
+  configuration options from long ago.
+- `git clone git@github.com:<your_name_here>/melos.git`
+- `git remote add upstream git@github.com:invertase/melos.git` (So that you
+  fetch from the main repository, not your clone, when running `git fetch` or `git pull`
+
+## Environment Setup
```
Comment (blaugold) [inline CONTRIBUTING.md:58]: This heading is not necessary.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+# ---------------------------------------------------------
+```
+
+### Install
+
+```bash
+dart pub global activate melos
+```
+
+## Performing changes
```
Comment (blaugold) [inline CONTRIBUTING.md:85]: This should be a level 3 heading.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+- Create a new local branch from `main` (e.g. `git checkout -b my-new-feature`)
+- Make your changes (try to split them up with one PR per feature/fix).
+- When committing your changes, make sure that each commit message is clear
+ (e.g. `git commit -m 'Added CONTRIBUTING.md'`).
```
Comment (blaugold) [inline CONTRIBUTING.md:90]: The example commit message should follow the conventional commit spec, e.g. `docs: Added CONTRIBUTING.md`.

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+- Push your new branch to your own fork into the same remote branch
+ (e.g. `git push origin my-username.my-new-feature`, replace `origin` if you use another remote.)
+
+ ### Open a pull request
```
Comment (blaugold) [inline CONTRIBUTING.md:94]: Please remove the leading space.

File: CONTRIBUTING.md (hunk 13 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:83]: This was suggested to be added [here](https://github.com/invertase/melos/issues/394#issuecomment-1270748150)

File: CONTRIBUTING.md (hunk 10 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:17]: Yes, I thought this was a part of the above point.

File: CONTRIBUTING.md (hunk 13 above)
Comment (blaugold) [reply CONTRIBUTING.md:80]: Yeah, that's a bit confusing. The `Install` section should be above the `Local development setup` section in the REAMDE.

Comment (blaugold) [reply CONTRIBUTING.md:80]: But in `CONTRIBUTING.md` we don't need the `Install` section.

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+### New Features
+
+- If you'd like to add a feature to the library that doesn't already exist, feel free to describe the feature in a new [GitHub issue](https://github.com/invertase/melos/issues/new?assignees=&labels=bug%2Ctriage&template=bug_report.yml&title=fix%3A++).
+- If you'd like to implement the new feature, please wait for feedback from the project maintainers before spending too much time writing the code. In some cases, enhancements may not align well with the project future development direction.
+- If applicable, implement the code for the new feature and please read the [How To](#how-to-contribute).
+
+### Documentation & Miscellaneous
+
+- If you have suggestions for improvements to the documentation or examples (or something else), we would love to hear about it.
+- As always first file a [Github issue].
```
Comment (nilsreichardt) [inline CONTRIBUTING.md:35]:
```suggestion
- As always first file a [GitHub issue](https://github.com/invertase/melos/issues/new).
```

@blaugold There is no issue template for this kind of issue, right?

**Review by nilsreichardt — COMMENTED** [review]
@blaugold Is there any formatting style for `.md` files?

File: CONTRIBUTING.md (hunk 18 above)
Comment (blaugold) [reply CONTRIBUTING.md:35]: That's correct. If you want to add one... :)

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+### Bug Fix
+
+- If you'd like to submit a fix for a bug, please read the [How To](#how-to-contribute) for how to send a Pull Request.
+- Indicate on the open issue that you are working on fixing the bug and the issue will be assigned to you.
+- Write `Fix #xxxx` in your PR text, where xxxx is the issue number (if there is one).
+- Include a test that isolates the bug and verifies that it was fixed.
+
+### New Features
+
+- If you'd like to add a feature to the library that doesn't already exist, feel free to describe the feature in a new [GitHub issue](https://github.com/invertase/melos/issues/new?assignees=&labels=bug%2Ctriage&template=bug_report.yml&title=fix%3A++).
```
Comment (blaugold) [inline CONTRIBUTING.md:28]:
```suggestion
- If you'd like to add a feature to the library that doesn't already exist, feel free to describe the feature in a new [GitHub issue](https://github.com/invertase/melos/issues/new?assignees=&labels=feature+request%2Ctriage&template=feature_request.yml&title=feature%3A++).
```

**Comment by blaugold** [comment]
> Is there any formatting style for .md files?

This project does not enforce one, but I like to use prettier for that.

File: CONTRIBUTING.md (hunk 13 above)
Comment (bhattabhi013) [reply CONTRIBUTING.md:61]: Removed.

Comment (bhattabhi013) [reply CONTRIBUTING.md:61]: > Yeah, that's a bit confusing. The `Install` section should be above the `Local development setup` section in the REAMDE.

Should I create a separate issue for it?

Comment (blaugold) [reply CONTRIBUTING.md:61]: You can just make that change in this PR. Maybe also replace `Local development setup` with `How to contribute` and link to this new document.

Comment (bhattabhi013) [reply CONTRIBUTING.md:61]: Then I shouldn't remove `Local development setup` from `CONTRIBUTING.md`.

**Review by blaugold — APPROVED** [review]

**Comment by blaugold** [comment]
Thank you for the contribution @bhattabhi013!

**Pull request merged**